### PR TITLE
mgr/dashboard: enable/disable versioning on RGW bucket

### PR DESCRIPTION
**Pull request by alfonsomthd**
Ability to enable/suspend bucket versioning.

UI design following proposal:
bucket-versioning-3

Fixes: https://tracker.ceph.com/issues/40920
Signed-off-by: alfonsomthd <almartin@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Comment by alfonsomthd** [comment]
jenkins test dashboard backend

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Comment by alfonsomthd** [comment]
jenkins test dashboard backend

**Comment by s0nea** [comment]
The last QA run failed: http://pulpito.ceph.com/tdehler-2019-08-06_06:33:36-rados:mgr-wip-tdehler-testing-29460-distro-basic-mira/

```
2019-08-07T04:32:20.022 INFO:tasks.cephfs_test_runner:======================================================================
2019-08-07T04:32:20.023 INFO:tasks.cephfs_test_runner:FAIL: test_get (tasks.mgr.dashboard.test_mgr_module.MgrModuleTelemetryTest)
2019-08-07T04:32:20.023 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-08-07T04:32:20.023 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-08-07T04:32:20.023 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-tdehler-testing-29460/qa/tasks/mgr/dashboard/test_mgr_module.py", line 121, in test_get
2019-08-07T04:32:20.023 INFO:tasks.cephfs_test_runner:    'license': str
2019-08-07T04:32:20.023 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-tdehler-testing-29460/qa/tasks/mgr/dashboard/helper.py", line 337, in assertSchema
2019-08-07T04:32:20.024 INFO:tasks.cephfs_test_runner:    self.assertEqual(data, str(e))
2019-08-07T04:32:20.024 INFO:tasks.cephfs_test_runner:AssertionError: {u'channel_crash': True, u'channel_basic': True, u'description': u'', u'url': u'https://telemetry.ceph.com/report', u'enabled': False, u'interval': 24, u'channel_device': True, u'channel_ident': False, u'contact': u'', u'proxy': u'', u'organization': u'', u'leaderboard': False} != "In `input`: missing keys: set(['license', 'report_timestamp', 'channels', 'report_version', 'channels_available', 'report_id'])"
```
Could it be related to https://tracker.ceph.com/issues/41055 / https://github.com/ceph/ceph/pull/29461 @tchaikov ?

**Comment by tchaikov** [comment]
@s0nea ack. just fixed my PR.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts [hunk 1]
```diff
@@ -124,19 +125,22 @@ export class RgwBucketFormComponent implements OnInit {
...
-        }
-      );
+      const versioningCtl = this.bucketForm.get('versioning');
+      this.rgwBucketService
+        .update(bidCtl.value, idCtl.value, ownerCtl.value, versioningCtl.value)
+        .subscribe(
+          () => {
+            this.notificationService.show(
+              NotificationType.success,
+              this.i18n('Updated Object Gateway bucket "{{bid}}"', { bid: bidCtl.value })
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts]: I'm not sure if we should close the sentence with a dot in a notification. Is there any guideline for that?
```suggestion
              this.i18n('Updated Object Gateway bucket "{{bid}}".', { bid: bidCtl.value })
```

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts]: Although the dot wasn't there even before this PR, I'll put it for consistency.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 2]
```diff
@@ -127,6 +127,43 @@
...
+                   id="enabled"
+                   name="versioning"
+                   formControlName="versioning"
+                   value="Enabled"
+                   [checked]="bucketForm.get('versioning').value == 'Enabled'"
+                   class="custom-control custom-radio custom-control-inline align-top">
+            <label class="align-text-top"
+                   for="enabled">
+              <span i18n>Enabled</span>
+              <div class="text-muted" i18n>Enables versioning for the objects in the bucket</div>
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Done.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 3]
```diff
@@ -127,6 +127,43 @@
...
+                   id="suspended"
+                   name="versioning"
+                   formControlName="versioning"
+                   value="Suspended"
+                   [checked]="bucketForm.get('versioning').value != 'Enabled'"
+                   class="custom-control custom-radio custom-control-inline align-top">
+            <label class="align-text-top"
+                   for="suspended">
+              <span i18n>Suspended</span>
+              <div class="text-muted" i18n>Disables versioning for the objects in the bucket</div>
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Done.

**Pull request merged**